staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -333,54 +334,86 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 		return r.base.RoundTrip(req)
 	}
 
-	creds, err := r.a.getCreds()
+	creds, _, err := r.a.getCreds()
 	if err != nil {
 		return nil, fmt.Errorf("getting credentials: %v", err)
 	}
 	if creds.token != "" {
+		req = utilnet.CloneRequest(req)
 		req.Header.Set("Authorization", "Bearer "+creds.token)
 	}
 
+	// make it possible for r.a.cert to detect when creds have already been fetched
+	ctx := context.WithValue(req.Context(), credsKey, creds)
+	req = req.WithContext(ctx)
+
 	res, err := r.base.RoundTrip(req)
 	if err != nil {
 		return nil, err
 	}
 	if res.StatusCode == http.StatusUnauthorized {
 		if err := r.a.maybeRefreshCreds(creds); err != nil {
-			klog.Errorf("refreshing credentials: %v", err)
+			klog.ErrorS(err, "refreshing credentials")
 		}
 	}
 	return res, nil
 }
 
-func (a *Authenticator) credsExpired() bool {
+// contextKey type is unexported to prevent collisions.
+type contextKey int
+
+const credsKey contextKey = iota
+
+func (a *Authenticator) credsExpiredLocked() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
-func (a *Authenticator) cert() (*tls.Certificate, error) {
-	creds, err := a.getCreds()
+func (a *Authenticator) cert(ctx context.Context) (*tls.Certificate, error) {
```

Comment:
since this is now returning errors in scenarios that only make sense when used as a `GetCert` callback, document this function is only used for that purpose (and maybe rename to `getCertCallback` or something to make it clear to callers?)